### Imports

In [13]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules
import random

## 1. Simulating Transactional Data

In [14]:
# Item list
items = ['Bread', 'Milk', 'Eggs', 'Cheese', 'Steak', 'Yogurt', 'Apples', 'Bananas']

# Generating 10 transactions with 2 to 5 items each(i set seed to make it more reproducible)
random.seed(142) 
transactions = []
for _ in range(10):
    num_items = random.randint(2, 5)
    transaction = random.sample(items, num_items)
    transactions.append(transaction)
    
for i, transaction in enumerate(transactions, 1):
    print(f"Transaction {i}: {transaction}")


Transaction 1: ['Eggs', 'Cheese', 'Bananas', 'Yogurt', 'Steak']
Transaction 2: ['Bread', 'Bananas', 'Eggs', 'Milk']
Transaction 3: ['Bread', 'Cheese', 'Milk']
Transaction 4: ['Yogurt', 'Steak']
Transaction 5: ['Milk', 'Bread', 'Cheese', 'Apples']
Transaction 6: ['Eggs', 'Yogurt', 'Bread', 'Cheese']
Transaction 7: ['Cheese', 'Yogurt', 'Eggs', 'Bananas', 'Bread']
Transaction 8: ['Steak', 'Yogurt', 'Eggs', 'Bread']
Transaction 9: ['Cheese', 'Steak']
Transaction 10: ['Yogurt', 'Bananas', 'Bread']


## 2. Analyze with Apriori

In [15]:
# ONe hot encode
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df = pd.DataFrame(te_ary, columns=te.columns_)
# Display the DataFrame
df.head()

Apples  Bananas  Bread  Cheese   Eggs   Milk  Steak  Yogurt
0   False     True  False    True   True  False   True    True
1   False     True   True   False   True   True  False   False
2   False    False   True    True  False   True  False   False
3   False    False  False   False  False  False   True    True
4    True    False   True    True  False   True  False   False

In [16]:
# Finding frequent Itemsets
frequent_itemsets = apriori(df, min_support=0.3, use_colnames=True)
print(frequent_itemsets)

    support                itemsets
0       0.4               (Bananas)
1       0.7                 (Bread)
2       0.6                (Cheese)
3       0.5                  (Eggs)
4       0.3                  (Milk)
5       0.4                 (Steak)
6       0.6                (Yogurt)
7       0.3        (Bread, Bananas)
8       0.3         (Eggs, Bananas)
9       0.3       (Yogurt, Bananas)
10      0.4         (Bread, Cheese)
11      0.4           (Bread, Eggs)
12      0.3           (Milk, Bread)
13      0.4         (Yogurt, Bread)
14      0.3          (Eggs, Cheese)
15      0.3        (Yogurt, Cheese)
16      0.4          (Yogurt, Eggs)
17      0.3         (Yogurt, Steak)
18      0.3   (Yogurt, Bread, Eggs)
19      0.3  (Yogurt, Eggs, Cheese)


## 3. Gemnerating Rules
Generating association Rules with minimum confidence of 0.7



In [17]:
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)

print("\nAssociation Rules:")
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])



Association Rules:
         antecedents consequents  support  confidence      lift
0          (Bananas)     (Bread)      0.3        0.75  1.071429
1          (Bananas)      (Eggs)      0.3        0.75  1.500000
2          (Bananas)    (Yogurt)      0.3        0.75  1.250000
3             (Eggs)     (Bread)      0.4        0.80  1.142857
4             (Milk)     (Bread)      0.3        1.00  1.428571
5             (Eggs)    (Yogurt)      0.4        0.80  1.333333
6            (Steak)    (Yogurt)      0.3        0.75  1.250000
7    (Yogurt, Bread)      (Eggs)      0.3        0.75  1.500000
8     (Yogurt, Eggs)     (Bread)      0.3        0.75  1.071429
9      (Bread, Eggs)    (Yogurt)      0.3        0.75  1.250000
10    (Yogurt, Eggs)    (Cheese)      0.3        0.75  1.250000
11  (Yogurt, Cheese)      (Eggs)      0.3        1.00  2.000000
12    (Cheese, Eggs)    (Yogurt)      0.3        1.00  1.666667


### Rule Explanation:
- If someone buys ['Bananas'], they are also likely to buy ['Bread']
- This rule has a confidence of 0.75, meaning that 75% of transactions containing the antecedent also contain the consequent.
- The support is 0.30, meaning these items appear together in 30% of all transactions.